backpack/custom_module/graph_utils.py

```diff
@@ -0,0 +1,203 @@
+"""Transformation tools to make graph BackPACK compatible."""
+from copy import deepcopy
+from typing import Set
+
+from torch.fx import Graph, GraphModule, Node, Tracer
+from torch.nn import Flatten, Module
+
+from backpack.custom_module.branching import ActiveIdentity, Branch, SumModule
+from backpack.custom_module.scale_module import ScaleModule
+from backpack.utils import TORCH_VERSION_AT_LEAST_1_9_0
+
+
+class BackpackTracer(Tracer):
+    """Tracer that recognizes BackPACK's custom modules as 'leaf modules'."""
+
+    def is_leaf_module(
+        self, m: Module, module_qualified_name: str
+    ) -> bool:  # noqa: D102
+        if isinstance(m, (ScaleModule, SumModule, Branch, ActiveIdentity)):
+            return True
+        else:
+            return super().is_leaf_module(m, module_qualified_name)
+
+
+def print_table(module: Module) -> None:
+    """Prints a table of the module.
+
+    Args:
+        module: module to analyze
+    """
+    graph_resnet18 = BackpackTracer().trace(module)
+    graph_resnet18.print_tabular()
+
+
+def convert_module_to_backpack(module: Module) -> GraphModule:
+    """Convert all modules to BackPACK-compatible modules.
+
+    Transformations:
+    - mul -> ScaleModule
+    - add -> AddModule
+    - flatten -> nn.Flatten
+
+    Args:
+        module: module to convert
+
+    Returns:
+        backpack compatible module
+
+    Raises:
+        NotImplementedError: if not torch >= 1.9.0
+    """
+    if TORCH_VERSION_AT_LEAST_1_9_0 is False:
+        raise NotImplementedError(
+            "Conversion is only possible for torch >= 1.9.0. This is because these "
+            "functions use functionality such as torch.nn.Module.get_submodule"
+        )
+    print("\nMake module BackPACK compatible...")
+    module_new = _transform_mul_to_scale_module(module)
+    module_new = _transform_flatten_to_module(module_new)
+    module_new = _transform_add_to_sum_module(module_new)
+    module_new = _transform_inplace_to_normal(module_new)
+    module_new = _transform_remove_duplicates(module_new)
+    print("\tDelete unused modules.")
+    module_new.delete_all_unused_submodules()
+    print("Finished transformation.\n")
+    return module_new
```

Comment:
This block should only `print` if a `debug` argument is set to `True` by the user. This function is called by [`extend`](https://github.com/tjs111/backpack/blob/bf5f132c9b8b1019b04e3887deb001f56f90d3ea/backpack/__init__.py#L221), and the value of `extend`'s `debug` argument can be passed to handle printing in the API.

Reply:
Added an option `debug: bool` (without default value). It is called from `extend` with the same value as `extend`.
I think it would be nice to have the tests using `debug=True`. But doesn't belong in this PR.